File: 백준/Silver/4963. 섬의 개수/섬의 개수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int w, h, count;
    static int[][] graph;
    static boolean[][] visited;
    static int[] dx = {-1, -1, -1, 0, 0, 1, 1, 1};  // x(행) 이동 방향
    static int[] dy = {-1, 0, 1, -1, 1, -1, 0, 1};  // y(열) 이동 방향


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        while (true) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            w = Integer.parseInt(st.nextToken()); // 너비
            h = Integer.parseInt(st.nextToken()); // 높이

            if (w == 0 && h == 0) {
                break;
            }

            graph = new int[h][w];
            for (int i = 0; i < h; i++) {
                st = new StringTokenizer(br.readLine());
                for (int j = 0; j < w; j++) {
                    graph[i][j] = Integer.parseInt(st.nextToken());
                }
            }

            count = 0;
            visited = new boolean[h][w];
            for (int i = 0; i < h; i++) {
                for (int j = 0; j < w; j++) {
                    if (!visited[i][j] && graph[i][j] == 1) {
                        count++;
                        dfs(i, j);
                    }
                }
            }
            sb.append(count).append("\n");
        }
        System.out.println(sb);

    }

    private static void dfs(int x, int y) {
        for (int i = 0; i < 8; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (0 <= nx && nx < h && 0 <= ny && ny < w) {
                if (!visited[nx][ny] && graph[nx][ny] == 1) {
                    visited[nx][ny] = true;
                    dfs(nx, ny);
                }
            }
        }
    }


}

File: 백준/Silver/4963. 섬의 개수/섬의 개수.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)


d = [(-1, -1), (-1, 0), (-1, 1),  # 북서, 북, 북동 
     (0, -1), (0, 1),  # 서, 동
     (1, -1), (1, 0), (1, 1)]  # 남서, 남, 남동


def dfs(x, y):
    for i in range(8):
        nx = x + d[i][0]
        ny = y + d[i][1]
        if 0 <= nx < h and 0 <= ny < w:
            if not visited[nx][ny] and graph[nx][ny] == 1:
                visited[nx][ny] = True
                dfs(nx, ny)


while True:
    w, h = map(int, input().split())  # w : 너비, h : 높이
    if w == h == 0:
        break

    graph = [[] for _ in range(h)]
    for i in range(h):
        graph[i] = list(map(int, input().split()))
    # print(graph)

    visited = [[False] * w for _ in range(h)]
    # print(visited)

    count = 0

    for i in range(h):
        for j in range(w):
            if not visited[i][j] and graph[i][j] == 1:
                count += 1
                dfs(i, j)
    print(count)